code/notificationbuilder/src/main/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/ProductCatalogNotificationBuilder.kt

```
@@ -0,0 +1,335 @@
+                PendingIntentUtils.createPendingIntent(
+                    context,
+                    trackerActivityClass,
+                    catalogItems[pushTemplate.currentIndex].uri,
+                    PushTemplateConstants.CatalogActionIds.PRODUCT_IMAGE_CLICKED,
+                    pushTemplate.tag,
+                    pushTemplate.isNotificationSticky ?: false
+                )
+            )
+        }
+        return downloadedImageCount
+    }
+
+    /**
+     * Sets the product catalog thumbnails and their click actions.
+     *
+     * @param context the application [Context]
+     * @param broadcastReceiverClass the [Class] of the broadcast receiver to set in the created pending intent
+     * @param channelIdToUse [String] containing the notification channel ID
+     * @param expandedLayout the [RemoteViews] object containing the expanded layout of the push template notification
+     * @param catalogItems the list of [ProductCatalogPushTemplate.CatalogItem] objects containing the product catalog items
+     * @param pushTemplate the [ProductCatalogPushTemplate] object containing the product catalog push template data
+     */
+    private fun populateThumbnails(
+        context: Context,
+        broadcastReceiverClass: Class<out BroadcastReceiver>?,
+        channelIdToUse: String,
+        expandedLayout: RemoteViews,
+        catalogItems: List<ProductCatalogPushTemplate.CatalogItem>,
+        pushTemplate: ProductCatalogPushTemplate
+    ) {
+        Log.trace(
+            LOG_TAG,
+            SELF_TAG,
+            "Populating product catalog thumbnails."
+        )
+
+        val thumbIds = listOf(
+            R.id.product_thumbnail_1,
+            R.id.product_thumbnail_2,
+            R.id.product_thumbnail_3
+        )
+        for (index in catalogItems.indices) {
+            val thumbImage = PushTemplateImageUtils.getCachedImage(catalogItems[index].img)
+
+            if (thumbImage == null) {
+                Log.trace(
```

Should we continue to construct the notification in this case or throw a `NotificationConstructionFailedException`

---

we should throw an exception as there must be 3 images. i'll update this.